tianshou/env/worker/subproc.py

```diff
@@ -0,0 +1,204 @@
+                if obs_bufs is not None:
+                    obs = _encode_obs(obs, obs_bufs)
+                p.send((obs, reward, done, info))
+            elif cmd == 'reset':
+                obs = env.reset()
+                if obs_bufs is not None:
+                    obs = _encode_obs(obs, obs_bufs)
+                p.send(obs)
+            elif cmd == 'close':
+                p.send(env.close())
+                p.close()
+                break
+            elif cmd == 'render':
+                p.send(env.render(**data) if hasattr(env, 'render') else None)
+            elif cmd == 'seed':
+                p.send(env.seed(data) if hasattr(env, 'seed') else None)
+            elif cmd == 'getattr':
+                p.send(getattr(env, data) if hasattr(env, data) else None)
+            else:
+                p.close()
+                raise NotImplementedError
+    except KeyboardInterrupt:
+        p.close()
+
+
+_NP_TO_CT = {
+    np.bool: ctypes.c_bool,
+    np.bool_: ctypes.c_bool,
+    np.uint8: ctypes.c_uint8,
+    np.uint16: ctypes.c_uint16,
+    np.uint32: ctypes.c_uint32,
+    np.uint64: ctypes.c_uint64,
+    np.int8: ctypes.c_int8,
+    np.int16: ctypes.c_int16,
+    np.int32: ctypes.c_int32,
+    np.int64: ctypes.c_int64,
+    np.float32: ctypes.c_float,
+    np.float64: ctypes.c_double,
+}
+
+
+class ShArray:
+    """Wrapper of multiprocessing Array"""
+
+    def __init__(self, dtype, shape):
+        self.arr = Array(_NP_TO_CT[dtype.type], int(np.prod(shape)))
+        self.dtype = dtype
+        self.shape = shape
+
+    def save(self, ndarray):
+        assert isinstance(ndarray, np.ndarray)
+        dst = self.arr.get_obj()
+        dst_np = np.frombuffer(dst, dtype=self.dtype).reshape(self.shape)
+        np.copyto(dst_np, ndarray)
+
+    def get(self):
+        return np.frombuffer(self.arr.get_obj(),
+                             dtype=self.dtype).reshape(self.shape)
+
+
+class SubprocEnvWorker(EnvWorker):
+    """Subprocess worker used in SubprocVectorEnv and ShmemVectorEnv."""
+
+    def __init__(self, env_fn: Callable[[], gym.Env],
+                 share_memory=False) -> None:
+        super().__init__(env_fn)
+        self.parent_remote, self.child_remote = Pipe()
+        self.share_memory = share_memory
+        self.buffer = None
+        if self.share_memory:
+            dummy = env_fn()
+            obs_space = dummy.observation_space
+            dummy.close()
+            del dummy
+            self.buffer = SubprocEnvWorker._setup_buf(obs_space)
+        args = (self.parent_remote, self.child_remote,
+                CloudpickleWrapper(env_fn), self.buffer)
+        self.process = Process(target=_worker, args=args, daemon=True)
+        self.process.start()
+        self.child_remote.close()
+
+    def __getattr__(self, key: str):
+        self.parent_remote.send(['getattr', key])
+        return self.parent_remote.recv()
+
+    @staticmethod
+    def _setup_buf(space):
```

Comment:
What it is point of this method ? It is meaningful to make it accessible to the user or it can only be used internally ? If the later, then imho it would not be able a nested method of `__init__`.